### PR TITLE
Implement various interfaces for trace configuration

### PR DESCRIPTION
**Pull request by tmplt**
I'm working on tracing support and aim to implement functions that abstract the configuration of relevant peripherals. Of chief interest is `DWT`, `ITM` and `TPIU`. Some propored abstractions will go against what is established in the crate; I will ask for comments on these.

### CODE REVIEW

File: src/peripheral/dwt.rs [hunk 1]
```diff
@@ -88,3 +116,147 @@ impl DWT {
...
+#[derive(Debug)]
+pub enum DWTError {
+    /// Invalid combination of [AccessType] and [EmitOption].
+    InvalidFunction,
+}
+
+impl Comparator {
+    /// Configure the function of the comparator
+    #[inline]
+    pub fn configure(&mut self, settings: ComparatorFunction) -> Result<(), DWTError> {
```
Comment (tmplt) [inline src/peripheral/dwt.rs:177]: Here is some API experimentation. Implementation is initial, and for now I've only implemented the option to configure for address comparison. Example usage:
```rust
    use cortex_m::peripheral as Core;
    use cortex_m::peripheral::{
        dwt::{AccessType, ComparatorAddressSettings, ComparatorFunction, EmitOption},
    };

    pub fn assign_dwt_unit(mut dwt: Core::dwt::Comparator) {
        let watch_address: u32 = unsafe { &super::WATCH_VARIABLE as *const _ } as u32;
        // TODO do we need to clear the MATCHED, bit[24] after every match?
        dwt.configure(ComparatorFunction::Address(ComparatorAddressSettings {
            address: watch_address,
            mask: 0,
            emit: EmitOption::Data,
            access_type: AccessType::WriteOnly,
        }))
        .unwrap(); // NOTE safe: valid (emit, access_type) used
    }
```

Thoughts? The `ComparatorFunction::Address(CompareAddressSettings {})` nest isn't very nice. But I'm unsure how to streamline it.

File: src/peripheral/tpiu.rs [hunk 2]
```diff
@@ -29,3 +31,71 @@ pub struct RegisterBlock {
...
+    /// The minimum implemented FIFO queue size of the TPIU for trace data.
+    min_queue_size: u8,
+}
+
+impl TPIU {
+    /// Sets the prescaler value for a wanted baud rate of the Serial
+    /// Wire Output (SWO) in relation to a given asynchronous refernce
+    /// clock rate.
+    #[inline]
+    pub fn set_swo_baud_rate(&mut self, ref_clk_rate: u32, baud_rate: u32) {
```
Comment (tmplt) [inline src/peripheral/tpiu.rs:63]: Is it possible to find the frequency of the asynchronous reference clock?

File: src/peripheral/dwt.rs [hunk 3]
```diff
@@ -70,6 +70,34 @@ impl DWT {
         unsafe { self.ctrl.modify(|r| r | 1) }
     }
 
+    /// Whether to enable exception tracing
+    // TODO find out if this is supported om armv6m
+    #[inline]
+    pub fn enable_exception_tracing(&mut self, bit: bool) {
```
Comment (tmplt) [inline src/peripheral/dwt.rs:76]: Should an enable argument be taken or should two `enable_exception_tracing`, `disable_exception_tracing` function be exposed instead, alike the API for DCB?

File: src/peripheral/tpiu.rs (hunk 2 above)
Comment (tmplt) [reply src/peripheral/tpiu.rs:63]: For the STM32F401RE, in RM0368, on p. 834:
> By default, the TRACECLKIN input clock of the TPIU is tied to GND. It is assigned to HCLK
two clock cycles after the bit TRACE_IOEN has been set.

Finding the frequency is thus assumed to be platform/vendor-specific. Nothing that can be done in this crate. Refer to HALs.

File: src/peripheral/dwt.rs (hunk 1 above)
Comment (tmplt) [reply src/peripheral/dwt.rs]: Implementation is now also ready for review, assuming that API changes are green-lit.

**Comment by tmplt** [comment]
API changes and implementations are ready for review. Some TODOs about armv6m remain. I'll resolve them after review.

**Comment by adamgreig** [comment]
Hi, just a quick note to say I just saw this via the reference in probe-rs and will try and review it for you soon, it would be great to have more support for trace management in here.

**Comment by tmplt** [comment]
We probably want to offer sane defaults for the added configuration structs.

File: src/peripheral/dwt.rs [hunk 4]
```diff
@@ -58,16 +69,57 @@ pub struct Comparator {
...
+        unsafe {
+            self.ctrl.modify(|mut r| {
+                r.set_cyccntena(true);
+                r
+            });
+        }
+    }
+
+    /// Whether to enable exception tracing
+    // TODO find out if this is supported om armv6m
```
Comment (adamgreig) [inline src/peripheral/dwt.rs:102]: I don't believe this is supported on ARMv6M: the architecture reference manual says of the DWT "This provides watchpoint support. Trace is not supported." and indeed there's no EXCTRCENA bit in DWT_CTRL.

File: src/peripheral/dwt.rs [hunk 5]
```diff
@@ -88,3 +140,112 @@ impl DWT {
...
+                {
+                    return Err(DWTError::InvalidFunction);
+                }
+
+                self.function.modify(|mut r| {
+                    // don't compare data value
+                    r.set_datavmatch(false);
+
+                    // dont compare cycle counter value
+                    // NOTE: only needed forp comparator 0, but is SBZP.
```
Comment (adamgreig) [inline src/peripheral/dwt.rs:214]:
```suggestion
                    // don;t compare cycle counter value
                    // NOTE: only needed for comparator 0, but is SBZP.
```

File: src/peripheral/itm.rs [hunk 6]
```diff
@@ -20,14 +23,30 @@ pub struct RegisterBlock {
...
+    /// Trace Control Register.
+    #[repr(C)]
+    #[derive(Copy, Clone)]
+    pub struct Tcr(u32);
+    get_itmena, set_itmena: 0;
+    get_tsena, set_tsena: 1;
+    get_syncena, set_synena: 2;
+    get_txena, set_txena: 3;
+    get_swoena, set_swoena: 4;
+    u8, get_tsprescale, set_tsprecale: 9, 8;
```
Comment (adamgreig) [inline src/peripheral/itm.rs:44]:
```suggestion
    u8, get_tsprescale, set_tsprescale: 9, 8;
```

**Review by adamgreig — CHANGES_REQUESTED** [review]
Thanks, sorry this sat waiting for so long!

On the whole this looks good to me. I've made a couple of comments. Besides that, could you fix up the clippy lints, which you can see in the PR view here? Mostly just needs some `#[allow]` on the capitalised names and a few other minor things.

File: src/peripheral/dwt.rs (hunk 4 above)
Comment (tmplt) [reply src/peripheral/dwt.rs]: Both functions gated from `armv6m`.

**Comment by tmplt** [comment]
Fixed. I don't have time to fix the merge conflict at the moment.

**Comment by tmplt** [comment]
CC @newAM, in case you want to refactor the code before I'm available.

**Comment by tmplt** [comment]
ping @adamgreig

File: src/peripheral/dwt.rs [hunk 7]
```diff
@@ -266,3 +302,112 @@ impl DWT {
...
+    pub mask: u32,
+    /// What sequence of packet(s) to emit on comparator match.
+    pub emit: EmitOption,
+    /// Whether to match on read, write or read/write operations.
+    pub access_type: AccessType,
+}
+
+/// The available functions of a DWT comparator.
+#[derive(Debug)]
+pub enum ComparatorFunction {
```
Comment (newAM) [inline src/peripheral/dwt.rs:348]: Mark as `#[non_exhaustive]` until the instruction/data/cycle count functions are implemented.

**Review by newAM — CHANGES_REQUESTED** [review]
I am not a maintainer, so take my review as suggestions from a fellow user :)

I double checked the register layouts and enumerated values, those all look good to me.

**Comment by tmplt** [comment]
Thanks for the review, @newAM. I'll go through it the coming weeks, if not this one.

File: src/peripheral/dwt.rs [hunk 8]
```diff
@@ -266,3 +302,112 @@ impl DWT {
...
+/// Possible error values returned on [Comparator::configure].
+#[derive(Debug)]
+pub enum DWTError {
+    /// Invalid combination of [AccessType] and [EmitOption].
+    InvalidFunction,
+}
+
+impl Comparator {
+    /// Configure the function of the comparator
+    #[inline]
```
[the remark this answers is not in the document]
Comment (tmplt) [reply src/peripheral/dwt.rs]: You're right. Fixed.

File: src/peripheral/itm.rs [hunk 9]
```diff
@@ -69,3 +88,106 @@ impl Stim {
...
+    SystemClock,
+    /// Clock timestamp counters using the asynchronous clock from the
+    /// TPIU interface.
+    ///
+    /// NOTE: The timestamp counter is held in reset while the output
+    /// line is idle.
+    AsyncTPIU,
+}
+
+/// blah
```
[the remark this answers is not in the document]
Comment (tmplt) [reply src/peripheral/itm.rs:136]: Oops, good catch.

File: src/peripheral/tpiu.rs [hunk 10]
```diff
@@ -19,13 +22,99 @@ pub struct RegisterBlock {
...
+    #[derive(Clone, Copy)]
+    pub struct Ffcr(u32);
+    get_enfcont, set_enfcont: 1;
+}
+
+bitfield! {
+    /// Type Register.
+    #[repr(C)]
+    #[derive(Clone, Copy)]
+    pub struct Type(u32);
```
[the remark this answers is not in the document]
Comment (tmplt) [reply src/peripheral/tpiu.rs:48]: It's the `TPIU_TYPE` register. C1.10.6 in https://developer.arm.com/documentation/ddi0403/ee/.

[the remark this answers is not in the document]
Comment (tmplt) [reply src/peripheral/tpiu.rs:48]: Field documentation has been updated in d62ee1a.

File: src/peripheral/tpiu.rs [hunk 11]
```diff
@@ -19,13 +22,99 @@ pub struct RegisterBlock {
...
+    u8, fifosz, _: 8, 6;
+    ptinvalid, _: 9;
+    mancvalid, _: 10;
+    nrzvalid, _: 11;
+}
+
+/// The available protocols for the trace output.
+pub enum TraceProtocol {
```
[the remark this answers is not in the document]
Comment (tmplt) [reply src/peripheral/tpiu.rs:65]: I refactored `SPPR` read/write to use bitfield in 5a3755f. A two-bit field is read/written, so it's `#[repr(u8)]` now.

File: src/peripheral/tpiu.rs [hunk 12]
```diff
@@ -16,16 +19,141 @@ pub struct RegisterBlock {
...
+    type Error = ();
+
+    /// Tries to convert from a `TXMODE` field value. Fails if the set mode is
+    /// unknown (and thus unpredictable).
+    #[inline]
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0b00 => Ok(Self::Parallel),
+            0b01 => Ok(Self::AsyncSWOManchester),
+            0b10 => Ok(Self::AsyncSWONRZ),
```
Comment (tmplt) [inline src/peripheral/tpiu.rs:83]: Unsure how values derived from the enum itself can be used.

File: src/peripheral/tpiu.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (tmplt) [reply src/peripheral/tpiu.rs:66]: As for common traits, I was not aware of the guideline. Fixed in 5bac9f5.

**Comment by tmplt** [comment]
Ready for a second review, @newAM. (Github's re-request functionality doesn't work on my end.)

File: src/peripheral/tpiu.rs (hunk 12 above)
Comment (newAM) [reply src/peripheral/tpiu.rs:84]: There is a way, but it isn't great.
```rs
    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            x if x == Self::Parallel as u8 => Ok(Self::Parallel),
            x if x == Self::AsyncSWOManchester as u8 => Ok(Self::AsyncSWOManchester),
            x if x == Self::AsyncSWONRZ as u8 => Ok(Self::AsyncSWONRZ),
            _ => Err(()), // unknown and unpredictable mode
        }
    }
```

Comment (tmplt) [reply src/peripheral/tpiu.rs:84]: Indeed. But I'll change to that just to that the mapping information mustn't be maintained in two places.

**Review by newAM — APPROVED** [review]
Looks good to me, thanks for all your work on this, I'm looking forward to using it :+1:

**Comment by jonathanpallant** [comment]
This looks OK to me. @newAM are you happy with the changes that arrived after your last review?

**Review by newAM — DISMISSED** [review]
Looks good to me!

File: src/peripheral/dwt.rs (hunk 3 above)
Comment (adamgreig) [reply src/peripheral/dwt.rs:159]: I think either have `set_exception_tracing(&mut self, enabled: bool)` _and/or_ `enable_exception_tracing(&mut self)`+`disable_exception_tracing(&mut self)`, but not `enable_exception_tracing(&mut self, bit: bool)`.

Comment (tmplt) [reply src/peripheral/dwt.rs]: Refactored into disable/enable functions.

**Comment by adamgreig** [comment]
Thanks! I think finally the MSRV needs to be bumped to support `non_exhaustive`, from 1.38 to 1.40 at least.

**Comment by adamgreig** [comment]
Thanks! I meant to say, it needs updating:

* in .github/workflows/ci.yml as you've done
* in .github/bors.toml so bors knows to wait for the new MSRV job
* in README.md and src/lib.rs for user-facing documentation

**Review by adamgreig — APPROVED** [review]
Perfect, thanks!

bors merge

**Pull request merged**